### ISSUE TITLE
passing current time to stepper for time dependent stimulation

### ISSUE THREAD

**Issue by Ziaeemehr** [ins-amu/vbjax#76]
There are some conditions that dfun need to apply a time dependent process on the system of equations, for example applying time dependent stimulation with start and end  time period, can we do this now or we need to pass time to stepper?

something like this in `loops.py`  [here](https://github.com/ins-amu/vbjax/blob/b4f27ec073d56eb85c0825fbc20687da64a39d45/vbjax/loops.py#L12)
```python
def heun_step(x, dfun, dt, t, *args, add=zero, adhoc=None, return_euler=False):
```
cheers,
A

**Comment by maedoc** [comment]
Thanks for opening this issue, this has been requested also by @mbreyt & @i-Zaak so let's step through some considerations in the implementation:

There are a few ways to do it; one is to go fully non-autonomous, so that the differential equations take time as an argument,
```python
def dfun(t: float, x: ndarray, p) -> ndarray:
   stim = np.sin(t * 2*np.pi * p.stim_freq)
   pass
```
another is to have time as an extra state variable, with a derivative of 1,
```python
def dfun(x, p):
    t, ... = x
    stim = np.sin(t * 2*np.pi * p.stim_freq)
    return np.array([1, ...]
```
but in the case where the time dependent process has already been evaluated and is just a an array in memory of shape `(n_time, ...)`, it might not make sense to interpolate the process in time, rather ensure that the time stepping occurs on the same grid as the time dep process itself (or some suitable ratio dt), then Jax would iterate this as an extra scan arg,
```python
def dfun(x, p, stim):
    pass

def op(c, xs):
    t, stim = xs
    return heun(x, dfun, dt, stim)

ts = np.r_[:n_time]
stim = np.sin(ts * 2*np.pi * p.stim_freq)
_, xs = jax.lax.scan(op, x, (ts, stim))
```
unfortunately, function signatures are not extensible, so this gets a bit complex over time.   One way to work around this is to employ a more flexible data structure which is then used in the dfun, 
```python
sim = {
    'time': 0.0,
    'stim': np.zeros((n_time, n_node), ),
    'x': x_init
    'param': ...
}

sim['stim'] = ...

ode_step, _ = vb.make_ode(...)

def sim_step(sim, t):
    sim['time'] = t
    sim['stim'] = np.sin(t*2*pi*sim['param'].stim_freq)
    sim['x'] = ode_step(sim['x'], sim['param'])
    return sim, sim['x']
   
sim, xs = jax.lax.scan(sim_step, sim, np.r_[:n_time])
```
This allows more customization of the details of the simulation without increasing the argument list for many functions like `dfun`, `heun` and `make_ode`, but retains benefits of Jax like JIT parallelization for GPU & gradients:
```
def run_one_freq(freq):
    sim['param'].stim_freq = freq
    sim, xs = jax.lax.scan(sim_step, sim, np.r_[:n_time])
    return xs

run_sweep = jax.jit(jax.vmap(run_one_freq))
run_grad = jax.jit(jax.grad(run_one_freq))
```
We would want to use actual [dataclasses](https://pypi.org/project/jax-dataclasses/) instead of dict to keep it easier to use but that's the idea. 

Can you @Ziaeemehr @i-Zaak @mbreyt comment on these ideas?

**Comment by Ziaeemehr** [comment]
The idea with a dictionary seems more customizable, I haven't worked with `dataclasses`, so if that's easier to use why not?
Extra state variable also seems good as a temporal solution.
Thanks for the explanation.

**Comment by i-Zaak** [comment]
Few showers later, here are some more thoughts.

I'm starting to like the integrated time state variable more, as I think it can be approached the same way as the multi-NMM simulations.

Making the timer an explicit "domain" I think helps to reason about this.

```python
def timer_dfun(t):
	return 1

def stim(t, p): # just an auxillary function, not integrated
    stim = np.sin(t * 2*np.pi * p.stim_freq)
    return stim
    
def dfun(tx,  p):
    t, x = tx # we would do this also for the franken-dfun
    st = stim(t, p.stim) # param name-spaces to avoid collisions?
   
    # reshape stim to x, apply differential weights based on nodes, stvars etc.
   
    xd = nmm_dfun(x, [c, st], p.nmm) # or c+stim, depends on the nmm...
    td = timer_dfun(t, p)
    
    return np.array([td, xd]) # shouldn't this be a list?
```

Alternatively, when the stimulus is additive to the nmm `dfun`, one would do something like this

```python
def dfun(tx,  p):
    t, x = tx # we would do this also for the franken-dfun
    st = stim(t, p.stim) 
    
    xd = nmm_dfun(x, c, p.nmm)
    td = timer_dfun(t, p)
    
    return np.array([td, xd + st]) 
```

Also the case when stimulus is precomputed can be implemented as a closure of stim over dt with rounded division inside, that I think is not an issue. 

Thoughts? :)

**Comment by maedoc** [comment]
Thanks for unpacking that idea a bit.  Since there's endless things that could go into this, and it's best to keep function args simple and unchanging for the API, I think I'd want to lean towards a dataclass approach
```
class State:
    step: int
    t: float  # just step*dt
    x: states
    ...

def dfun(state: State, p: Param) -> Deriv:
    ...
```
which is a little more obvious than tuples of tuples, and `jax.jit` makes this free.
